osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs

```diff
@@ -14,51 +16,97 @@ namespace osu.Game.Rulesets.Osu.Difficulty.Skills
+            {
+                if (osuCurrObj.Angle != null)
                 {
-                    const double scale = 90;
+                    double angle = osuCurrObj.Angle.Value;
+
+                    // Rewarding angles, take the smaller velocity as base.
+                    double angleBonus = Math.Min(currVector.Length, prevVector.Length);
 
-                    var angleBonus = Math.Sqrt(
-                        Math.Max(osuPrevious.JumpDistance - scale, 0)
-                        * Math.Pow(Math.Sin(osuCurrent.Angle.Value - angle_bonus_begin), 2)
-                        * Math.Max(osuCurrent.JumpDistance - scale, 0));
-                    result = 1.4 * applyDiminishingExp(Math.Max(0, angleBonus)) / Math.Max(timing_threshold, osuPrevious.StrainTime);
+                    double wideAngleBonus = calcWideAngleBonus(angle);
+                    double acuteAngleBonus = calcAcuteAngleBonus(angle);
+
+                    if (osuCurrObj.StrainTime > 100)
+                        acuteAngleBonus = 0;
+                    else
+                    {
+                        acuteAngleBonus *= Math.Min(2, Math.Pow((100 - osuCurrObj.StrainTime) / 15, 1.5));
+                        wideAngleBonus *= Math.Pow(osuCurrObj.StrainTime / 100, 6);
+                    }
+
+                    if (acuteAngleBonus > wideAngleBonus)
+                        angleBonus = Math.Min(angleBonus, 150 / osuCurrObj.StrainTime) * Math.Min(1, Math.Pow(Math.Min(osuCurrObj.JumpDistance, osuPrevObj.JumpDistance) / 150, 2));
+
+                    angleBonus *= Math.Max(acuteAngleBonus * acute_angle_multiplier, wideAngleBonus * wide_angle_multiplier);
+
+                    // add in angle velocity.
+                    aimStrain += angleBonus;
                 }
             }
+            else // There is a rhythm change
+            {
+                // Rewarding rhythm, take the smaller velocity as base.
+                double rhythmBonus = Math.Min(currVector.Length, prevVector.Length);
 
-            double jumpDistanceExp = applyDiminishingExp(osuCurrent.JumpDistance);
-            double travelDistanceExp = applyDiminishingExp(osuCurrent.TravelDistance);
+                if (osuCurrObj.StrainTime + 10 < osuPrevObj.StrainTime && osuPrevObj.StrainTime > osuLastObj.StrainTime + 10)
+                    // Don't want to reward for a rhythm change back to back (unless its a double, which is why this only checks for fast -> slow -> fast).
+                    rhythmBonus = 0;
 
-            return Math.Max(
-                result + (jumpDistanceExp + travelDistanceExp + Math.Sqrt(travelDistanceExp * jumpDistanceExp)) / Math.Max(osuCurrent.StrainTime, timing_threshold),
-                (Math.Sqrt(travelDistanceExp * jumpDistanceExp) + jumpDistanceExp + travelDistanceExp) / osuCurrent.StrainTime
-            );
+                aimStrain += rhythmBonus * rhythm_variance_multiplier; // add in rhythm velocity.
```

Comment:
Honestly I'm not sure if aim should reward rhythm at all, this should be covered by rhythm complexity alone. Aiming rhythm changes isn't hard by itself, reading might be, but it's out of scope of aim skill